---
title: "Karite genetic parameters"
author: "Karen Cristine Goncalves dos Santos"
date: "`r Sys.Date()`"
output: 

  pdf_document:
    fig_caption: yes
    keep_tex: yes
---

```{r Load packages, include=FALSE}
packages_to_use = c(
	"poppr", "mmod", "adegenet", "ape", "LEA",
	"treemap", "tidyverse",
	"vcfR", "pegas", "hierfstat",
	"ggpubr")

# install.packages(packages_to_use,
# 								 dependencies = TRUE)
for (i in packages_to_use) { 
	library(i, character.only = T)}
```

```{r Load pop info, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
population_information <- read.csv(file = "input data - DO NOT MODIFY/Population.csv")
names(population_information) <- c("Taxa", "Taxa.Name", "Country",
																	 "Region", "Location.Population",
																	 "Type.Land.Use")

population_information = 
	population_information[- which(is.na(population_information$Taxa)),]

```

The file  here was processed with STACKS with the following filters:
#     min_cov: minimum allele coverage to keep genotype <int>: 3
#     percent_genotypes: minimum percent of genotype data per population <float> : 50
#     max_pop_fail: maximum number of populations that can fail percent_genotypes <int> : 0
#     min_mas: minimum number of samples with rare allele <int> : 2

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
vcf = read.vcfR("input data - DO NOT MODIFY/third_filters_m3_p50_x0_S2.singleton.unlinked_100k.vcf")
genind = vcfR2genind(vcf, sep = "/")
popinfo = read.delim("input data - DO NOT MODIFY/Population.txt", header = T)
popmap = unique(popinfo[, 2:5])

inds_info <- rownames(tab(genind))
sep_info <- data.frame(
  ind = inds_info,
  Country = gsub("(\\w{3}).+", "\\1", inds_info),
  Region = gsub("\\w{4}\\d-(\\w+)_.+", "\\1", inds_info),
  Locality = gsub("\\w{4}\\d-(\\w+)_.+", "\\1", inds_info),
  Land_use = gsub("\\w{4}\\d-\\w+_(\\w{2}).+", "\\1", inds_info),
  sample = gsub("\\w{4}\\d-\\w+_\\w{2}-pl\\d+-[A-H][0-9]+-(\\d+)", "\\1", inds_info)
)

true_names = list(
  Country = c("Cameroun", "Mali", "Tchad"),
  Region = c(rep("Extreme-North", 3), rep("North", 3),
             rep("Adamaoua", 3), rep("West", 4),
             "Segou", "Moyen-Chari", "Logone Occidental",
             "Mandoul", "Logone Oriental", "Mayo-kebbi", "Mandoul", 
             "Tandjile"),
  Locality = c("Mayo-Dana", "Mayo-Kani", "Mayo-Tsana", "Bascheo",
               "Ngong", "Tcheboa", "Falaise", "Tchabal", "Tison",
               "Bangoua", "Koutaba", "Mbouda", "Tonga", "Solosso",
               "Kagala", "Kana", "Kemkiam", "Mayongo", "Pala",
               "Peni", "Tchoua"),
  Land_use = c("Champ agricole", "Savane naturelle", 'Jardin de case',
               "Champ", "Jachere", "Verger", "Foret")
)

for (i in names(true_names)) {
  names(true_names[[i]]) = unique(sep_info[, i])
}

new_popmap = matrix(nrow = 0, ncol = 5) %>% as.data.frame()
names(new_popmap) = c("Sample", names(true_names))

for (i in 1:nrow(sep_info)) {
  ind_data = data.frame(
    Sample = sep_info$ind[i],
    Country = true_names[["Country"]][
      sep_info[i, "Country"]
    ],
    Region = true_names[["Region"]][
      sep_info[i, "Region"]
    ],
    Locality = true_names[["Locality"]][
      sep_info[i, "Locality"]
    ],
    Land_use = true_names[["Land_use"]][
      sep_info[i, "Land_use"]
    ])
  
  new_popmap = data.frame(
    rbind(new_popmap, ind_data)
  )
}
```

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
strata(genind) <- new_popmap
setPop(genind) <- ~Country/Region/Locality

save(genind, new_popmap, sep_info, file = "basic_genind.RData")
```

```{r}
rm(list=ls())
load("basic_genind.RData")
groups <- with(new_popmap, 
							 list(Country = unique(Country),
							 		 Region = unique(Region),
							 		 Locality = unique(Locality),
							 		 Land_use = unique(Land_use)))

```
```{r summary per country}
country_genind <- genind
setPop(country_genind) <-
	~Country

genetic_summary_countries <- 
	lapply(groups$Country, function(x) {
	inds <- pop(country_genind) == x
	
	basic.stats(country_genind[inds])
})

names(genetic_summary_countries) <- groups$Country



alle_rich_countries = allelic.richness(country_genind)$Ar %>%
	apply(2, mean)

priv_alle_countries = private_alleles(country_genind, report = "table", level = "population") %>%
	rowSums()

n_pop_countries = sapply(groups$Country, function(x) {
	length(which(new_popmap$Country == x))
	})

basic_stats_countries = lapply(groups$Country, function(x) {
	data.frame(
		N = n_pop_countries[x],
		Ho = genetic_summary_countries[[x]]$overall["Ho"],
		Hs = genetic_summary_countries[[x]]$overall["Hs"],
		Fis = genetic_summary_countries[[x]]$overall["Fis"],
		Private_alleles = priv_alle_countries[x],
		Allelic_richness = alle_rich_countries[x]
	)
	
}) %>% do.call(what = rbind)
```


```{r summary per pop}
genetic_summary_pop <- 
	lapply(levels(pop(genind)), function(x) {
	inds <- pop(genind) == x
	
	basic.stats(genind[inds])
})
names(genetic_summary_pop) <- levels(pop(genind))

n_pop = sapply(levels(pop(genind)), function(x) {
	length(which(pop(genind) == x))
	})

hexp = Hs(genind, pop = pop(genind))

alle_rich_pop = allelic.richness(genind)$Ar %>%
	apply(2, mean)

priv_alle_pop = private_alleles(genind, report = "table", level = "population") %>%
	rowSums()


basic_stats_pop= lapply(levels(pop(genind)), function(x) {
	data.frame(
		N = n_pop[x],
		Ho = genetic_summary_pop[[x]]$overall["Ho"],
		Hs = hexp[x],
		Fis = genetic_summary_pop[[x]]$overall["Fis"],
		Private_alleles = priv_alle_pop[x],
		Allelic_richness = alle_rich_pop[x] %>% round(digits = 3)
	)
	
}) %>% do.call(what = rbind)

write.table(basic_stats_pop, 
						file = paste0("BasicStats_pop_", Sys.Date(), ".txt"),
						row.names = T, col.names = T, 
						append = F, quote = F, sep = "\t", )
```

## Calculate heterozygosity

```{r Heterozygosity}
# Test homogeneity of variances of observed and expected heterozygosities
summaries = lapply(levels(pop(genind)), function(x) {
	inds = pop(genind) == x
	summary(genind[inds])
})
names(summaries) = levels(pop(genind))

homogeneity = lapply(levels(pop(genind)), function(x) {
	bartlett.test(list(summaries[[x]]$Hexp, 
										 summaries[[x]]$Hobs))
})
names(homogeneity) = levels(pop(genind))

# Test if expected het is greater than observed. 
# Tell the test if the variances are equal or not
zygosity = lapply(levels(pop(genind)), function(x) {
	
	var.equal = ifelse(homogeneity[[x]]$p.value > 0.05,
										 T, F)
	
	t.test(summaries[[x]]$Hexp, 
				 summaries[[x]]$Hobs,
				 pair = T, var.equal = var.equal, 
				 alternative = "greater",
				 conf.level = 0.95)
})
names(zygosity) = levels(pop(genind))

```

## Get global F-stats

```{r}
fst_fis = wc(genind)


fstats = Fst(genind %>% as.loci) %>% colMeans()

```

```{r Amova with Mali}
amova_mali <- poppr.amova(x = genind, within = F,
													hier = ~Country/Region/Locality,
													threads = parallel::detectCores(), cutoff = 0.95
)

(rand_amova_mali <- randtest(amova_mali, nrepet = 999))

plot(rand_amova_mali)

knitr::kable(amova_mali$results, caption = 
						 	"Result AMOVA including the Mali population.\
						 Hierarchy: _Country - Region - Locality_.\
						 Sample = Location.Population")

knitr::kable(amova_mali$componentsofcovariance, caption = 
						 	"Components of covariance from AMOVA including the Mali population.\
						 Hierarchy: _Country - Region - Locality_.\
						 Sample = Location.Population")
```


```{r Amova without Mali, eval=FALSE, include=FALSE}
exclude_mali <- population_information$inds[
	population_information$Country != "Mali"]

genind_nomali <- genind[exclude_mali]
strata(genind_nomali) <- population_information[
	population_information$Country != "Mali",] %>%
	apply(MARGIN = 2, FUN = as.character) %>% as.data.frame
setPop(genind_nomali) <-
	~Country/Region/Locality

amova_nomali <- poppr.amova(x = genind_nomali, within = F,
														hier = ~Country/Region/Locality,
														threads = parallel::detectCores(), cutoff = 0.95
)

(rand_amova_nomali <- randtest(amova_nomali, nrepet = 999))

plot(rand_amova_nomali)

knitr::kable(amova_nomali$results, caption = 
						 	"Result AMOVA excluding the Mali population.\
						 Hierarchy: _Country - Region - Locality_.\
						 Sample = Location.Population")

```

```{r glPCA, eval=FALSE, include=FALSE}
vcf = read.vcfR("input data - DO NOT MODIFY/third_filters_m3_p50_x0_S2.singleton.unlinked_100k.vcf")

genlight <- vcfR2genlight(vcf, n.cores = parallel::detectCores())
genlight@pop = pop(genind)
genlight@strata = strata(genind)
genlight@hierarchy = genind@hierarchy

glPCA_result <- glPca(
	genlight, parallel = T, nf = 20,
	n.cores = parallel::detectCores()
)

glPCA_df <- data.frame(
	Sample = rownames(glPCA_result$scores),
	glPCA_result$scores
) %>% merge(new_popmap, by = "Sample") 

#devtools::install_github("jaredhuling/jcolors")
library("jcolors")
theme_set(theme_bw())
ggplot(glPCA_df, 
			 aes(PC1, PC2,shape = Country, color = Region)) +
	geom_point() + scale_color_jcolors(palette = "pal8")
```

# DAPC

## snapclust
```{r find number of clusters, echo=FALSE}
aic <- snapclust.choose.k(21, genind, IC = "AIC")
plot(aic, type = "b", cex = 2, xlab = "k", ylab = "AIC")
points(which.min(aic), min(aic), col = "blue", pch = 20, cex = 3)
abline(v = 21, lty = 2, col = "red")

# #bic <- snapclust.choose.k(21, genind, IC = "BIC")
# plot(bic, type = "b", cex = 2, xlab = "k", ylab = "BIC")
# points(which.min(bic), min(bic), col = "blue", pch = 20, cex = 3)
# abline(v = 21, lty = 2, col = "red")

nk = which.min(aic)
```

## DAPC using clusters obtained with find.clusters function


```{r DAPC with find.clusters, echo=FALSE}
clusters_karite <- find.clusters(
	genind, n.pca = 490, stat = "AIC",
	method = "ward", n.clust = nk)

table(pop(genind), clusters_karite$grp)
											 
table.value(
	table(new_table$pop, clusters_karite$grp),
	col.labels = paste("K ", levels(clusters_karite$grp)))

dapc<- dapc.genind(genind,
									 clusters_karite$grp, 
									 n.pca = 50, n.da = 2
)

dapc_da <- data.frame(
	new_popmap[, 2:5],
	clusters = dapc$grp,
	dapc$ind.coord
)
```


```{r DAPC scatter plot, echo=FALSE, fig.height=10, fig.width=13, paged.print=FALSE}
library(gridExtra)
library(RColorBrewer)
library(pals)

theme_set(theme_bw())
scatter_dapc <- list(
	membership = (
		data.frame(table(new_table$pop, clusters_karite$grp)) %>% 
		 	mutate(new_freq = ifelse(Freq == 0, NA, Freq), 
		 				 Ks = factor(Var2, levels = 1:nk), 
		 				 pops = factor(Var1, levels = levels(pop(genind))[21:1])) %>%
		 	ggplot(aes(Ks, pops, size = new_freq)) + 
		 	geom_point(shape = 22, fill = "black") + 
		 	theme(axis.text.x.top = element_text()) + 
		 	labs(x = "Clusters", y = "") + 
		 	scale_size_continuous(name = "Number of individuals", 
		 												breaks = c(5, seq(10, 40, 10)), 
		 												limits = c(0, 45))
		 ),
	# eigen = (
	# 	data.frame(
	# 	DAs = factor(paste0("Discriminant function ", 1:dapc$n.da),
	# 							 levels = paste0("Discriminant function ", 1:dapc$n.da)),
	# 	eigenvalues = dapc$eig) %>%
	# 		ggplot(aes(DAs, eigenvalues)) +
	# 		geom_col() + 
	# 		labs(x = "", y = "Eigenvalues", 
	# 				 title = "Eigenvalues of retained discriminant functions") +
	# 		#scale_y_continuous(breaks = seq(0, 6000, 1000)) +
	# 		theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
	# 					legend.position = "none")
	# ), 
	ld1_ld2 =
		(ggplot(dapc_da, aes(LD1, LD2, color = clusters)) + 
		 	geom_point(aes(shape = Country)) + 
		 	stat_ellipse(aes(color = clusters)) +
		 	labs(x = "DAPC LD1", y = "DAPC LD2") +
		 	scale_color_manual(values = pals::glasbey(nk), name = "Clusters"))
)

ggsave(
   filename = paste0("ReferenceSNP_Scatters_DAPC_findclusters", Sys.Date(), ".pdf"), 
   plot = marrangeGrob(scatter_dapc, nrow=1, ncol=1)
)

```

```{r DAPC PCA plot}
data.frame(inds = new_popmap$Sample, 
           grp = dapc$grp, 
           Country = new_popmap$Country,
           PC1 = dapc$tab$`PCA-pc.1`, 
           PC2 = dapc$tab$`PCA-pc.2`) %>% 
ggplot(aes(PC1, PC2, color = grp, shape = Country)) + geom_point() +
    scale_color_manual(values = pals::glasbey(nk))


ggsave(
   filename = paste0("ReferenceSNP_PCA_DAPC_findclusters", Sys.Date(), ".pdf")
)

```
Structure-like plot - findclusters doesnt work because the posterior probability of each individual is 1 in all cases.

```{r Structure-like plot - findclusters, eval=FALSE, fig.height=7.21, fig.width=13.9, include=FALSE}
membership_probability = data.frame(
	individuals = rownames(dapc_countries[["posterior"]]),
	dapc_countries[["posterior"]]
) %>% merge(
	data.frame(individuals = new_popmap$Sample,
						 pops = as.character(pop(genind))),
	by = "individuals") %>%
	pivot_longer(
	-c(individuals, pops), names_to = "Clusters", 
	values_to = "probability", names_prefix = "X")

membership_probability$individuals <- 
	factor(membership_probability$individuals,
				 levels = new_popmap$Sample[
				 	order(new_popmap$Sample)])

mem_prob_pops <- lapply(levels(pop(genind)), function(i) {
		title = paste0(
			i, "\nN=", which(pop(genind) == i) %>% length
		)
		filter(membership_probability, pops == i) %>%
		ggplot(
       aes(individuals, probability, fill = Clusters)) +
    geom_col() + labs(x = "", y = "",
    									title = title) +
    scale_fill_manual(values = glasbey(nK), breaks = 1:nK) +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
    			axis.title.y = element_text(vjust = 0.5),
          panel.grid = element_blank(), 
    			plot.title = element_text(hjust = 0.5)
    			)
	})

plot_groups <- list(
	ggarrange(plotlist = lapply(1:6, function(i) {mem_prob_pops[[i]]}),
					ncol = 2, nrow = 3, labels = "AUTO", 
					common.legend = T, legend = "bottom"),
	ggarrange(plotlist = lapply(7:12, function(i) {mem_prob_pops[[i]]}),
					ncol = 2, nrow = 3, labels = "AUTO", 
					common.legend = T, legend = "bottom"),
	ggarrange(plotlist = lapply(13:18, function(i) {mem_prob_pops[[i]]}),
					ncol = 2, nrow = 3, labels = "AUTO", 
					common.legend = T, legend = "bottom"),
	ggarrange(plotlist = lapply(19:21, function(i) {mem_prob_pops[[i]]}),
					ncol = 1, nrow = 3, labels = "AUTO", 
					common.legend = T, legend = "bottom"))
library(gridExtra)

ggsave(
	filename = paste0("ReferenceSNP_Cluster_membership_DAPC_DA10_PC90_k", nK, Sys.Date(), ".pdf"), 
   plot = marrangeGrob(plot_groups, nrow=1, ncol=1)
)

```



```{r Pairwise Fst - calculate it and save, eval=FALSE, include=FALSE}
pwfst <- genet.dist(genind, diploid = T, method = "Fst")
save(pwfst, file = "RData/GenetDist_Fst.RData")
```


```{r Pairwise Fst}
load("RData/GenetDist_Fst.RData")
nj_fst <- pwfst %>% nj()
nj_fst$edge.length <- abs(nj_fst$edge.length)
nj_fst$tip.label <- 
	levels(pop(genind)) %>%
	gsub(pattern = "_", replacement = " - ")

pwfst_df <- as.matrix(pwfst) %>% 
	as.data.frame()
names(pwfst_df) <- rownames(pwfst_df) <- 
	levels(pop(genind))

dend_pwfst <- hclust(pwfst_df %>% as.matrix %>% as.dist)
pops_countries <- dend_pwfst$labels %>% 
	str_split(pattern = "_") %>% 
	lapply(function(x) x[1]) %>%
	unlist %>% as.factor
```


```{r Pairwise Fst dend plot}
library(ggdendro)

ggdend_pwfst <- as.dendrogram(dend_pwfst) %>%
	dendro_data()

ggdend_pwfst$labels$label <- 
	gsub("_", " - ", ggdend_pwfst$labels$label)

ggdend_pwfst$labels$country <- 
	ggdend_pwfst$labels$label %>% str_split(pattern = " ") %>%
	lapply(function(x) x[1]) %>% unlist %>% as.factor 

ggplot() +
    geom_segment(data = ggdend_pwfst$segments, 
                 aes(x = y, y = x, 
                     xend = yend, yend = xend))+
    geom_text(data = ggdend_pwfst$labels, 
              aes(y, x, label = label, color = country),
              hjust = 1.1, size = 3) +
    scale_x_continuous(
    	limits = c(-0.3, 0.3), 
    	breaks = seq(0, 0.3, 0.05)) + 
	theme_minimal() + 
	theme(axis.text.y = element_blank(), 
				panel.grid = element_blank(), 
				legend.position = "none") + 
	labs(x = "Pairwise Fst", y = "")

ggsave(paste0("Reference_snps_PairwiseFST_njtree.pdf"))
knitr::kable(pwfst_df, digits = 3)
```



